File: cpp/color_detector.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <iostream>
#include <map>

//using namespace std;
//using namespace cv;


std::map<std::string, std::tuple<cv::Scalar_<double>, cv::Scalar_<double>>> colours = {
  {"Rojo", std::make_tuple(cv::Scalar(155,25,0), cv::Scalar(179,255,255))},
  {"Azul", std::make_tuple(cv::Scalar(100,200,50), cv::Scalar(140,255,255))}
};

//colours = {"Rojo": (Scalar(155,25,0), Scalar(179,255,255)),
//           "Azul": (Scalar(100,200,200), Scalar(140,255,255))
//           }

cv::Mat colorFilter(const cv::Mat& src, std::tuple<cv::Scalar_<double>, cv::Scalar_<double>> hsvInterval)
{
  assert(src.type() == CV_8UC3);
  cv::Mat colour;
  inRange(src, std::get<0>(hsvInterval), std::get<1>(hsvInterval), colour);
  return colour;
}


int main( int argc, char** argv )
{
  cv::VideoCapture cap(0); //capture the video from web cam

  if ( !cap.isOpened() )  // if not success, exit program
  {
    std::cout << "Cannot open the web cam" << std::endl;
    return -1;
  }
   //inRange(imgHSV, Scalar(iLowH, iLowS, iLowV), Scalar(iHighH, iHighS, iHighV), imgThresholded); //Threshold the image

  while (true)
  {
    cv::Mat imgOriginal;

    bool bSuccess = cap.read(imgOriginal); // read a new frame from video

    if (!bSuccess) //if not success, break loop
    {
      std::cout << "Cannot read a frame from video stream" << std::endl;
      break;
    }

    cv::Mat imgHSV;

    cv::cvtColor(imgOriginal, imgHSV, cv::COLOR_BGR2HSV); //Convert the captured frame from BGR to HSV

    cv::Mat imgThresholded;

    std::map<std::string, std::tuple<cv::Scalar_<double>, cv::Scalar_<double>>>::iterator it = colours.begin();
    for (std::pair<std::string, std::tuple<cv::Scalar_<double>, cv::Scalar_<double>>> element : colours) {

      imgThresholded = colorFilter(imgHSV, element.second);

      int numWhitePixels = cv::countNonZero(imgThresholded);
      if (numWhitePixels > 10000) {
        std::cout << element.first << ": "<< numWhitePixels << '\n';

        cv::Mat result;
        cv::bitwise_and(imgOriginal, imgOriginal, result, imgThresholded);

        cv::imshow("Result Image", result); //show the thresholded image
        cv::imshow("Original", imgOriginal); //show the original image
      }
    }

    //morphological closing (fill small holes in the foreground)
    //cv::dilate(imgThresholded, imgThresholded, cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5)));
    //cv::erode(imgThresholded, imgThresholded, cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5)));

    if (cv::waitKey(30) == 27) //wait for 'esc' key press for 30ms. If 'esc' key is pressed, break loop
    {
      std::cout << "esc key is pressed by user" << std::endl;
      break;
    }
  }
  return 0;
}

File: python/color_detector.py
import cv2
import numpy as np

cap = cv2.VideoCapture(0)

width = cap.get(3)  # float
height = cap.get(4)  # float
total_size = width * height

colours = {"Rojo": ([125,100,0], [179,255,255]),
           "Azul": ([100,200,50], [140,255,255]),
           "Verde": ([65,55,90], [90,255,255])
           #"Amarillo": ([103, 86, 65], [145, 133, 128])
           }
while 1:
    _, frame = cap.read()
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    #Get HSV interval for mask
    #col_min = np.uint8([[[0, 0, 70]]])
    #col_max = np.uint8([[[170, 100, 255]]])
    #hsv_col_min = cv2.cvtColor(col_min,cv2.COLOR_BGR2HSV)
    #hsv_col_max = cv2.cvtColor(col_max,cv2.COLOR_BGR2HSV)
    #print (hsv_col_min, hsv_col_max)

    #lower = np.array([125,100,0])
    #upper = np.array([179,255,255])

    for colour in colours:
        lower = np.array(colours[colour][0])
        upper = np.array(colours[colour][1])

        mask = cv2.inRange(hsv, lower, upper)
        img_dilated = cv2.dilate(mask, np.ones((8,8), np.uint8), iterations=1)
        res = cv2.bitwise_and(frame, frame, mask=img_dilated)

        contours, hierarchy = cv2.findContours(img_dilated, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)

        cv2.drawContours(frame, contours, -1, (0, 255, 0), 3)
        n_white_pix = np.sum(mask == 255)

        #print(colour, n_white_pix)
        if n_white_pix > 10000:
            print("Red", n_white_pix)

        cv2.imshow('mask', mask)
        cv2.imshow('morph', img_dilated)
        cv2.imshow('res', res)
        cv2.imshow('frame', frame)

    k = cv2.waitKey(5) & 0xFF
    if k == 27:
        break

cv2.destroyAllWindows()
cap.release()
